### PR TITLE
Aleph alpha llm

### PR DESCRIPTION
**Pull request by ulan-yisaev**
# Description

This PR introduces the integration of [Aleph Alpha](https://app.aleph-alpha.com/)'s API into the llama-index, enabling users to leverage Aleph Alpha's Large Language Models (LLMs) and embedding capabilities within the llama-index's retrieval ecosystem.

## New Package?

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [x] Yes
- [ ] No

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [ ] Yes
- [x] No

## Type of Change

- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.

- [x] Added new unit/integration tests
- [x] Added new notebook (that tests end-to-end)
- Documentation on how to run these tests is provided

## Suggested Checklist:

- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] I ran `make format; make lint` to appease the lint gods.

### CODE REVIEW

File: llama-index-integrations/embeddings/llama-index-embeddings-alephalpha/README.md
Comment (nerdai) [inline llama-index-integrations/embeddings/llama-index-embeddings-alephalpha/README.md:1]: Could we fill in some details for this README.md? This appears in our llama-hub and is the first piece of information that a visitor to llama-hub will see on this package.

File: llama-index-integrations/llms/llama-index-llms-alephalpha/llama_index/llms/alephalpha/BUILD
Comment (nerdai) [inline llama-index-integrations/llms/llama-index-llms-alephalpha/llama_index/llms/alephalpha/BUILD:1]: (Same comment as for embeddings)
Could we fill in some details for this README.md? This appears in our llama-hub and is the first piece of information that a visitor to llama-hub will see on this package.

File: llama-index-integrations/llms/llama-index-llms-alephalpha/llama_index/llms/alephalpha/base.py [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+        )
+
+        self.token = get_from_param_or_env("aa_token", token, "AA_TOKEN", "")
+
+        self._client = None
+        self._aclient = None
+
+    @classmethod
+    def class_name(cls) -> str:
+        return "AlephAlpha_LLM"
```
Comment (nerdai) [inline llama-index-integrations/llms/llama-index-llms-alephalpha/llama_index/llms/alephalpha/base.py:125]: nit: I think this should be `AlephAlpha`?

File: llama-index-integrations/llms/llama-index-llms-alephalpha/llama_index/llms/alephalpha/base.py [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+    @llm_completion_callback()
+    def complete(
+        self, prompt: str, formatted: bool = False, **kwargs: Any
+    ) -> CompletionResponse:
+        client = self._get_client()
+        all_kwargs = {"prompt": Prompt.from_text(prompt), **self._completion_kwargs}
+
+        request = CompletionRequest(**all_kwargs)
+
+        response = client.complete(request=request, model=self.model)
```
Comment (nerdai) [inline llama-index-integrations/llms/llama-index-llms-alephalpha/llama_index/llms/alephalpha/base.py:206]: out of curiosity: does this include `LogProbs` for next tokens?

File: llama-index-integrations/llms/llama-index-llms-alephalpha/llama_index/llms/alephalpha/base.py [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+        completion = response.completions[0].completion
+        return CompletionResponse(text=completion, raw=response.to_json())
+
+    @llm_completion_callback()
```
Comment (nerdai) [inline llama-index-integrations/llms/llama-index-llms-alephalpha/llama_index/llms/alephalpha/base.py:210]: just noting that, our `callbacks` will be replaced by our new `instrumentation`. We will support `callbacks` for a while still but just giving you a heads up that this mechanism will need to change at some point. :)

File: llama-index-integrations/llms/llama-index-llms-alephalpha/tests/test_llms_alephalpha.py
Comment (nerdai) [inline llama-index-integrations/llms/llama-index-llms-alephalpha/tests/test_llms_alephalpha.py:1]: feel free to add your own unit tests here as well. :)

File: docs/examples/llm/alephalpha.ipynb
Comment (nerdai) [inline docs/examples/llm/alephalpha.ipynb:1]: I think perhaps this is a mistake? This should be a notebook for LLMs, but instead it looks to be the same notebook as the one for embeddings?

**Review by nerdai — CHANGES_REQUESTED** [review]
Generally this looks pretty good @ulan-yisaev - thanks for the contribution!

I left a few comments though that need to get addressed before we can get this merged.

File: docs/examples/llm/alephalpha.ipynb
Comment (ulan-yisaev) [reply docs/examples/llm/alephalpha.ipynb:1]: sorry, yes, I accidentally replaced this notebook :) I'll fix this in the next commit.

File: llama-index-integrations/embeddings/llama-index-embeddings-alephalpha/README.md
Comment (ulan-yisaev) [reply llama-index-integrations/embeddings/llama-index-embeddings-alephalpha/README.md:1]: Sure thing :)

File: llama-index-integrations/llms/llama-index-llms-alephalpha/llama_index/llms/alephalpha/base.py (hunk 2 above)
Comment (ulan-yisaev) [reply llama-index-integrations/llms/llama-index-llms-alephalpha/llama_index/llms/alephalpha/base.py:206]: Yeah, I can integrate log_probs and other additional params

Comment (ulan-yisaev) [reply llama-index-integrations/llms/llama-index-llms-alephalpha/llama_index/llms/alephalpha/base.py:249]: I've added the `additional_kwargs` attribute to the response object. This attribute contains additional information such as log probabilities, model version, and raw completion.

**Review by nerdai — APPROVED** [review]
thanks for making the changes @ulan-yisaev - lgtm!

**Comment by nerdai** [comment]
merging this - thanks @ulan-yisaev for the contribution!

**Comment by ulan-yisaev** [comment]
Thank you, @nerdai, for your support and for merging the changes!

**Pull request merged**